File: jazzer/src/main/java/dev/erst/fingrind/jazzer/support/JazzerTopology.java
package dev.erst.fingrind.jazzer.support;

import dev.erst.fingrind.jazzer.tool.JazzerJson;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

/** Loads and validates the committed Jazzer topology model used by local tooling and wrappers. */
final class JazzerTopology {
  private static final String RESOURCE_PATH =
      "/dev/erst/fingrind/jazzer/support/jazzer-topology.json";
  private static final Registry REGISTRY = load();

  private JazzerTopology() {}

  static Registry registry() {
    return REGISTRY;
  }

  private static Registry load() {
    return loadResource(RESOURCE_PATH);
  }

  static Registry loadResource(String resourcePath) {
    Objects.requireNonNull(resourcePath, "resourcePath must not be null");
    TopologyDocument document;
    try {
      document = JazzerJson.readResource(resourcePath, TopologyDocument.class);
    } catch (IOException exception) {
      throw new UncheckedIOException("Failed to load Jazzer topology resource", exception);
    }
    return load(document);
  }

  static Registry load(TopologyDocument document) {
    List<JazzerHarness> harnesses = new ArrayList<>();
    Map<String, JazzerHarness> harnessesByKey = new ConcurrentHashMap<>();
    for (HarnessDocument harnessDocument : document.harnesses()) {
      JazzerHarness harness = toHarness(harnessDocument);
      if (harnessesByKey.putIfAbsent(harness.key(), harness) != null) {
        throw new IllegalArgumentException("Duplicate Jazzer harness key: " + harness.key());
      }
      harnesses.add(harness);
    }

    List<JazzerRunTarget> runTargets = new ArrayList<>();
    Map<String, JazzerRunTarget> runTargetsByKey = new ConcurrentHashMap<>();
    Map<String, JazzerRunTarget> runTargetsByTaskName = new ConcurrentHashMap<>();
    for (RunTargetDocument runTargetDocument : document.runTargets()) {
      JazzerRunTarget runTarget = toRunTarget(runTargetDocument, harnessesByKey);
      if (runTargetsByKey.putIfAbsent(runTarget.key(), runTarget) != null) {
        throw new IllegalArgumentException("Duplicate Jazzer run target key: " + runTarget.key());
      }
      if (runTargetsByTaskName.putIfAbsent(runTarget.taskName(), runTarget) != null) {
        throw new IllegalArgumentException("Duplicate Jazzer task name: " + runTarget.taskName());
      }
      validateActiveRunTarget(runTarget);
      runTargets.add(runTarget);
    }

    return new Registry(
        List.copyOf(harnesses),
        Map.copyOf(harnessesByKey),
        List.copyOf(runTargets),
        Map.copyOf(runTargetsByKey),
        Map.copyOf(runTargetsByTaskName));
  }

  private static JazzerHarness toHarness(HarnessDocument harnessDocument) {
    return new JazzerHarness(
        JazzerHarnessKind.fromKey(harnessDocument.key()),
        harnessDocument.displayName(),
        harnessDocument.className(),
        harnessDocument.methodName());
  }

  private static JazzerRunTarget toRunTarget(
      RunTargetDocument runTargetDocument, Map<String, JazzerHarness> harnessesByKey) {
    List<JazzerHarness> harnesses =
        runTargetDocument.harnessKeys().stream()
            .map(key -> requireHarness(harnessesByKey, key))
            .toList();
    return new JazzerRunTarget(
        runTargetDocument.key(),
        runTargetDocument.displayName(),
        runTargetDocument.taskName(),
        runTargetDocument.workingDirectory(),
        runTargetDocument.activeFuzzing(),
        harnesses);
  }

  private static JazzerHarness requireHarness(
      Map<String, JazzerHarness> harnessesByKey, String harnessKey) {
    JazzerHarness harness = harnessesByKey.get(harnessKey);
    if (harness == null) {
      throw new IllegalArgumentException("Unknown Jazzer harness key in topology: " + harnessKey);
    }
    return harness;
  }

  private static void validateActiveRunTarget(JazzerRunTarget runTarget) {
    if (!runTarget.activeFuzzing()) {
      return;
    }
    if (runTarget.harnesses().size() != 1) {
      throw new IllegalArgumentException(
          "Active Jazzer run target must reference exactly one harness: " + runTarget.key());
    }
    if (!runTarget.harnesses().getFirst().key().equals(runTarget.key())) {
      throw new IllegalArgumentException(
          "Active Jazzer run target key must match its harness key: " + runTarget.key());
    }
  }

  record Registry(
      List<JazzerHarness> harnesses,
      Map<String, JazzerHarness> harnessesByKey,
      List<JazzerRunTarget> runTargets,
      Map<String, JazzerRunTarget> runTargetsByKey,
      Map<String, JazzerRunTarget> runTargetsByTaskName) {
    Registry {
      harnesses = harnesses == null ? List.of() : List.copyOf(harnesses);
      harnessesByKey = harnessesByKey == null ? Map.of() : Map.copyOf(harnessesByKey);
      runTargets = runTargets == null ? List.of() : List.copyOf(runTargets);
      runTargetsByKey = runTargetsByKey == null ? Map.of() : Map.copyOf(runTargetsByKey);
      runTargetsByTaskName =
          runTargetsByTaskName == null ? Map.of() : Map.copyOf(runTargetsByTaskName);
    }
  }

  record TopologyDocument(List<HarnessDocument> harnesses, List<RunTargetDocument> runTargets) {
    TopologyDocument {
      harnesses = harnesses == null ? List.of() : List.copyOf(harnesses);
      runTargets = runTargets == null ? List.of() : List.copyOf(runTargets);
      if (harnesses.isEmpty()) {
        throw new IllegalArgumentException("Jazzer topology must declare at least one harness");
      }
      if (runTargets.isEmpty()) {
        throw new IllegalArgumentException("Jazzer topology must declare at least one run target");
      }
    }
  }

  record HarnessDocument(String key, String displayName, String className, String methodName) {
    HarnessDocument {
      key = requireNonBlank(key, "key");
      displayName = requireNonBlank(displayName, "displayName");
      className = requireNonBlank(className, "className");
      methodName = requireNonBlank(methodName, "methodName");
    }
  }

  record RunTargetDocument(
      String key,
      String displayName,
      String taskName,
      String workingDirectory,
      boolean activeFuzzing,
      List<String> harnessKeys) {
    RunTargetDocument {
      key = requireNonBlank(key, "key");
      displayName = requireNonBlank(displayName, "displayName");
      taskName = requireNonBlank(taskName, "taskName");
      workingDirectory = requireNonBlank(workingDirectory, "workingDirectory");
      harnessKeys = harnessKeys == null ? List.of() : List.copyOf(harnessKeys);
      if (harnessKeys.isEmpty()) {
        throw new IllegalArgumentException("harnessKeys must not be empty");
      }
      harnessKeys.forEach(harnessKey -> requireNonBlank(harnessKey, "harnessKey"));
    }
  }

  private static String requireNonBlank(String value, String fieldName) {
    Objects.requireNonNull(value, fieldName + " must not be null");
    String trimmed = value.trim();
    if (trimmed.isEmpty()) {
      throw new IllegalArgumentException(fieldName + " must not be blank");
    }
    return trimmed;
  }
}

File: gradle/build-logic/src/main/kotlin/dev/erst/fingrind/buildlogic/JazzerTopology.kt
package dev.erst.fingrind.buildlogic

import groovy.json.JsonSlurper
import org.gradle.api.Project

internal data class JazzerHarnessSpec(
    val key: String,
    val displayName: String,
    val className: String,
    val methodName: String,
) {
    fun inputResourceDirectory(): String {
        val packagePath = className.substringBeforeLast('.').replace('.', '/')
        val simpleName = className.substringAfterLast('.')
        return "$packagePath/${simpleName}Inputs/$methodName"
    }
}

internal data class JazzerRunTargetSpec(
    val key: String,
    val displayName: String,
    val taskName: String,
    val workingDirectory: String,
    val activeFuzzing: Boolean,
    val harnessKeys: List<String>,
)

internal data class JazzerTopology(
    val harnesses: List<JazzerHarnessSpec>,
    val runTargets: List<JazzerRunTargetSpec>,
) {
    private val harnessesByKey: Map<String, JazzerHarnessSpec> = harnesses.associateBy(JazzerHarnessSpec::key)
    private val runTargetsByKey: Map<String, JazzerRunTargetSpec> = runTargets.associateBy(JazzerRunTargetSpec::key)

    fun harness(key: String): JazzerHarnessSpec =
        harnessesByKey[key] ?: throw IllegalArgumentException("Unknown Jazzer harness key: $key")

    fun runTarget(key: String): JazzerRunTargetSpec =
        runTargetsByKey[key] ?: throw IllegalArgumentException("Unknown Jazzer run target key: $key")

    companion object {
        private const val HARNESSES_PATH = "src/main/resources/dev/erst/fingrind/jazzer/support/jazzer-harnesses.json"
        private const val RUN_TARGETS_PATH = "src/main/resources/dev/erst/fingrind/jazzer/support/jazzer-run-targets.json"

        fun load(project: Project): JazzerTopology {
            val harnessesFile = project.layout.projectDirectory.file(HARNESSES_PATH).asFile
            val runTargetsFile = project.layout.projectDirectory.file(RUN_TARGETS_PATH).asFile
            val harnessesRoot = JsonSlurper().parse(harnessesFile) as? List<*>
                ?: throw IllegalArgumentException("Jazzer harness catalog must be a JSON array: $HARNESSES_PATH")
            val runTargetsRoot = JsonSlurper().parse(runTargetsFile) as? List<*>
                ?: throw IllegalArgumentException("Jazzer run-target catalog must be a JSON array: $RUN_TARGETS_PATH")
            val harnesses =
                maps(harnessesRoot, "harnesses").map { map ->
                    JazzerHarnessSpec(
                        key = string(map, "key"),
                        displayName = string(map, "displayName"),
                        className = string(map, "className"),
                        methodName = string(map, "methodName"),
                    )
                }
            val runTargets =
                maps(runTargetsRoot, "runTargets").map { map ->
                    JazzerRunTargetSpec(
                        key = string(map, "key"),
                        displayName = string(map, "displayName"),
                        taskName = string(map, "taskName"),
                        workingDirectory = string(map, "workingDirectory"),
                        activeFuzzing = boolean(map, "activeFuzzing"),
                        harnessKeys = strings(map["harnessKeys"], "harnessKeys"),
                    )
                }
            require(harnesses.isNotEmpty()) { "Jazzer topology must declare at least one harness" }
            require(harnesses.map(JazzerHarnessSpec::key).toSet().size == harnesses.size) {
                "Jazzer topology harness keys must be unique"
            }
            require(runTargets.isNotEmpty()) { "Jazzer topology must declare at least one run target" }
            require(runTargets.map(JazzerRunTargetSpec::key).toSet().size == runTargets.size) {
                "Jazzer topology run target keys must be unique"
            }
            val topology = JazzerTopology(harnesses = harnesses, runTargets = runTargets)
            topology.validate()
            return topology
        }

        private fun string(map: Map<*, *>, key: String): String =
            map[key] as? String ?: throw IllegalArgumentException("Missing string field '$key' in Jazzer topology")

        private fun boolean(map: Map<*, *>, key: String): Boolean =
            map[key] as? Boolean ?: throw IllegalArgumentException("Missing boolean field '$key' in Jazzer topology")

        private fun maps(value: List<*>, key: String): List<Map<*, *>> =
            value.map { item ->
                item as? Map<*, *> ?: throw IllegalArgumentException("Field '$key' must contain JSON objects")
            }

        private fun strings(value: Any?, key: String): List<String> =
            (value as? List<*>)?.map { item ->
                item as? String ?: throw IllegalArgumentException("Field '$key' must contain strings")
            } ?: throw IllegalArgumentException("Missing array field '$key' in Jazzer topology")
    }

    private fun validate() {
        runTargets.forEach { target ->
            require(target.harnessKeys.isNotEmpty()) {
                "Jazzer run target '${target.key}' must reference at least one harness"
            }
            target.harnessKeys.forEach { harness(it) }
            if (target.activeFuzzing) {
                require(target.harnessKeys.size == 1) {
                    "Active Jazzer run target '${target.key}' must reference exactly one harness"
                }
                require(target.harnessKeys.single() == target.key) {
                    "Active Jazzer run target '${target.key}' must share its key with its harness"
                }
            }
        }
    }
}
